libgadget/petapm.h

```diff
@@ -16,7 +16,46 @@ typedef struct Region {
     int no; /* node number for debugging */
 } PetaPMRegion;
 
-typedef struct PetaPMPriv PetaPMPriv; /* private */
+/* a layout is the communication object, represent
+ * pencil / cells exchanged  */
+
+struct Layout {
+    MPI_Comm comm;
+    int NpExport;
+    int NpImport;
+    int * NpSend;
+    int * NpRecv;
+    int * DpSend;
+    int * DpRecv;
+    struct Pencil * PencilSend;
+    struct Pencil * PencilRecv;
+
+    int NcExport;
+    int NcImport;
+    int * NcSend;
+    int * NcRecv;
+    int * DcSend;
+    int * DcRecv;
+
+    double * BufSend;
+    double * BufRecv;
+    int * ibuffer;
+};
+
+/* Data which is private to the PetaPM structure. Don't access from outside.*/
+typedef struct PetaPMPriv {
```

Comment:
The reason of keeping a Priv is to avoid exposing unnecessary interface to the public file. I think the benefit of keeping a smaller API outweights saving an allocation -- and the obligation to maintain the life cycle of these variables the way it currently is (which is not very well documented.)

Unless there are other reasons a client must see all of the fields. If it is just one or two fields, we can always add the accessor functions.

Reply:
I agree that a small API is generally desirable, but we have no obligation to keep binary compatibility or do actual data member hiding in this code. We can keep the API small by just not using priv outside of petapm. Whereas the extra allocation is another thing to ensure gets freed in the right order, which is a bit of a pain. So while I agree with you as a general rule of code, in this specific case I think the stack allocation is the cleaner path.

Reply:
Sure.

On the other hand, priv can be managed by the standard C malloc. Only `large` and frequently realloced objects (fragment makers) really 'need' to use mymalloc. `priv` is unlikely a fragmenter in this sense.

I think when I was hacking in FastPM, there were several occasions I hoped an object was an opaque pointer. I don't quite recall the reason.  Probably related to forward declarations or embedding structures. I'll bring this up again if I remember the situation.